### ISSUE TITLE
bundle install fails with jruby 9.2.5.0 - NoMethodError: undefined method `[]' for nil:NilClass

### ISSUE THREAD

**Issue by Mogztter** [jruby/jruby#6008]
# Error Report
## Questions
Please fill out answers to these questions, it'll help us figure out
why things are going wrong.
- **What did you do?**
  I ran the command `/home/travis/.rvm/gems/jruby-9.2.5.0/bin/bundle install --jobs=3 --retry=3`
- **What did you expect to happen?**
  I expected Bundler to install the dependencies
- **What happened instead?**
  Instead, what happened was it throws an error!
- **Have you tried any solutions posted on similar issues in our issue tracker, stack overflow, or google?**
  Yes.
- **Have you read our issues document, https://github.com/bundler/bundler/blob/master/doc/contributing/ISSUES.md?**
  Yes.
## Backtrace
```
NoMethodError: undefined method `[]' for nil:NilClass
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/source/git/git_proxy.rb:90:in `version'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/source/git/git_proxy.rb:147:in `block in copy_to'
  org/jruby/RubyDir.java:399:in `chdir'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/shared_helpers.rb:52:in `block in chdir'
  /home/travis/.rvm/rubies/jruby-9.2.5.0/lib/ruby/stdlib/monitor.rb:226:in `mon_synchronize'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/shared_helpers.rb:51:in `chdir'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/source/git/git_proxy.rb:136:in `copy_to'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/source/git.rb:163:in `specs'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/definition.rb:764:in `block in converge_locked_specs'
  org/jruby/RubyArray.java:1792:in `each'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/definition.rb:750:in `converge_locked_specs'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/definition.rb:247:in `resolve'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/definition.rb:170:in `specs'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/definition.rb:158:in `resolve_remotely!'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/installer.rb:303:in `resolve_if_needed'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/installer.rb:84:in `block in run'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/process_lock.rb:12:in `block in lock'
  org/jruby/RubyIO.java:1154:in `open'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/process_lock.rb:9:in `lock'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/installer.rb:73:in `run'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/installer.rb:25:in `install'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/cli/install.rb:65:in `run'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/cli.rb:256:in `block in install'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/settings.rb:124:in `temporary'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/cli.rb:255:in `install'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/cli.rb:30:in `dispatch'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/cli.rb:24:in `start'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/exe/bundle:46:in `block in <main>'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'
  /home/travis/.rvm/gems/jruby-9.2.5.0/gems/bundler-2.1.2/exe/bundle:34:in `<main>'
  org/jruby/RubyKernel.java:1007:in `load'
  /home/travis/.rvm/gems/jruby-9.2.5.0/bin/bundle:1:in `(root)'
  org/jruby/RubyKernel.java:1046:in `eval'
  /home/travis/.rvm/gems/jruby-9.2.5.0/bin/jruby_executable_hooks:24:in `<main>'
```
## Environment
```
Bundler       2.1.2
  Platforms   ruby, universal-java-11
Ruby          2.5.0p0 (2018-12-06 revision 60928) [java]
  Full Path   /home/travis/.rvm/rubies/jruby-9.2.5.0/bin/jruby
  Config Dir  /home/travis/.rvm/rubies/jruby-9.2.5.0/etc
RubyGems      2.7.6
  Gem Home    /home/travis/.rvm/gems/jruby-9.2.5.0
  Gem Path    /home/travis/.rvm/gems/jruby-9.2.5.0:/home/travis/.rvm/gems/jruby-9.2.5.0@global
  User Home   /home/travis
  User Path   /home/travis/.gem/jruby/2.5.0
  Bin Dir     /home/travis/.rvm/gems/jruby-9.2.5.0/bin
OpenSSL       
  Compiled    JRuby-OpenSSL 0.10.2
  Loaded      JRuby-OpenSSL 0.10.2
  Cert File   /usr/local/lib/jvm/openjdk11/lib/security/cacerts
  Cert Dir    /etc/ssl/certs
Tools         
  Git         
  RVM         1.29.7 (latest)
  rbenv       rbenv 1.1.2
  chruby      not installed
```
## Bundler Build Metadata
```
Built At          2019-12-20
Git SHA           4da3289ebd
Released Version  true
```
## Bundler settings
```
retry
  Set for your local app (/home/travis/build/asciidoctor/asciidoctor-reveal.js/.bundle/config): 3
jobs
  Set for your local app (/home/travis/build/asciidoctor/asciidoctor-reveal.js/.bundle/config): 3
gemfile
  Set via BUNDLE_GEMFILE: "/home/travis/build/asciidoctor/asciidoctor-reveal.js/Gemfile"
```
### Gemfile
```ruby
source 'https://rubygems.org'
# Specify your gem's dependencies in asciidoctor-revealjs.gemspec
gemspec
# Asciidoctor.js 2.0.0 requires an unreleased Opal 0.11.99.dev (d136ea8)
gem 'opal', :git => 'https://github.com/opal/opal.git', :ref => 'd136ea8'
```
### Gemfile.lock
```
<No /home/travis/build/asciidoctor/asciidoctor-reveal.js/Gemfile.lock found>
```
### asciidoctor-revealjs.gemspec
```ruby
# -*- encoding: utf-8 -*-
require File.expand_path '../lib/asciidoctor-revealjs/version', __FILE__
require 'open3'
Gem::Specification.new do |s|
  s.name = 'asciidoctor-revealjs'
  s.version = Asciidoctor::Revealjs::VERSION
  s.authors = ['Olivier Bilodeau']
  s.email = ['olivier@bottomlesspit.org']
  s.homepage = 'https://github.com/asciidoctor/asciidoctor-reveal.js'
  s.summary = 'A reveal.js converter for Asciidoctor. Write your slides in AsciiDoc!'
  s.description = 'Converts AsciiDoc documents into HTML5 presentations designed to be executed by the reveal.js presentation framework.'
  s.license = 'MIT'
  files = begin
    if (result = Open3.popen3('git ls-files -z') {|_, out| out.read }.split %(\0)).empty?
      Dir['**/*']
    else
      # converter.rb is built locally before packaging but ignored by git. Adding manually.
      result + ['lib/asciidoctor-revealjs/converter.rb']
    end
  rescue
    Dir['**/*']
  end
  # TODO should we still package template files now that they are built into ruby?
  s.files = files.grep %r/^(?:(?:examples|lib|templates)\/.+|Gemfile|Rakefile|(?:CHANGELOG|LICENSE|README)\.adoc|#{s.name}\.gemspec)$/
  s.executables = ['asciidoctor-revealjs']
  s.extra_rdoc_files = Dir['README.adoc', 'LICENSE.adoc', 'HACKING.adoc']
  s.require_paths = ['lib']
  s.add_runtime_dependency 'asciidoctor', ['>= 2.0.0', '< 3.0.0']
  s.add_runtime_dependency 'thread_safe', '~> 0.3.5'
  s.add_runtime_dependency 'concurrent-ruby', '~> 1.0'
  s.add_development_dependency 'rake', '~> 10.4.2'
  s.add_development_dependency 'asciidoctor-doctest', '= 2.0.0.beta.5'
  s.add_development_dependency 'pry', '~> 0.10.4'
  if RUBY_ENGINE != 'jruby'
    s.add_development_dependency 'pry-byebug'
    s.add_development_dependency 'pygments.rb'
  end
  s.add_development_dependency 'colorize'
  s.add_development_dependency 'asciidoctor-templates-compiler', '~> 0.4.2'
  s.add_development_dependency 'slim', '~> 3.0.6'
  s.add_development_dependency 'slim-htag', '~> 0.1.0'
  s.add_development_dependency 'rouge'
  # Overriden in Gemfile and Gemfile.upstream for now
  #s.add_development_dependency 'opal', '~> 0.11.1'
end
```

Here's the link to the build on Travis: https://travis-ci.org/asciidoctor/asciidoctor-reveal.js/jobs/627564886

Please note that it's working with `jruby-head` (jruby 9.2.10.0-SNAPSHOT).

**Comment by Mogztter** [comment]
Even though this issue is already fixed on master/head I think it's important to "document" it because `jruby-9k` on Travis is still using 9.2.5.0 (and not the latest release [9.2.9.0](https://github.com/jruby/jruby/tree/9.2.9.0))

**Comment by kares** [comment]
might be a good place to put on the wiki, otherwise not sure what JRuby could do.
old releases just get more broken with newer libraries - and to document all of that ...
I am thinking this should "really" be raised at Travis's end?
for JRuby, probably having this issue here might help already users searching for a workaround.

**Comment by Mogztter** [comment]
I think I found the root cause. Travis is using rvm version 1.29.7 while the latest stable version available is 1.29.9.
The latest known version of jruby in rvm 1.29.7 is: `jruby-9.2.5.0`.

Travis should probably update the xenial image to use rvm version 1.29.9. Though I'm not sure where I should open an issue... https://travis-ci.community/t/upgrade-rvm-to-the-latest-stable-version/6644

To workaround this issue, I guess we need to manually update rvm:

```travis.yml
before_install:
- rvm get stable # or rvm get head
```

**Comment by Mogztter** [comment]
> To workaround this issue, I guess we need to manually update rvm:

It's actually does not work because `before_install` runs after the `rvm` matrix... :disappointed:

**Comment by dr-itz** [comment]
@Mogztter the easiest workaround is to specify `jruby-9.2.9.0` instead of `jruby-9k`. Older rvm doesn't "know" it but will fetch it anyways. We're doing the same in `activerecord-jdbc-adapter`. The downside is that you have to update from time to time as newer jruby versions get released

**Comment by Mogztter** [comment]
> the easiest workaround is to specify jruby-9.2.9.0 instead of jruby-9k. Older rvm doesn't "know" it but will fetch it anyways.

Yes that's what we are doing :+1: 

> The downside is that you have to update from time to time as newer jruby versions get released

Indeed, that's why I was looking for an alternative solution.

**Comment by headius** [comment]
Since this has already been fixed, I think this bug will serve as good documentation of the issue and how to work around it (upgrade JRuby). I will mark this as fixed as of JRuby 9.2.9.0.

**Comment by ramanamatya** [comment]
> Since this has already been fixed, I think this bug will serve as good documentation of the issue and how to work around it (upgrade JRuby). I will mark this as fixed as of JRuby 9.2.9.0.

fix where ?

**Comment by enebo** [comment]
@ramanamatya any place you can get a version of JRuby 9.2.9.0+.  RVM, our website jruby.org, other ruby managers.
